### PR TITLE
Engine example to load .drc and .ply format using loaders.gl

### PR DESCRIPTION
**Pull request by tonetechnician**
Fixes: #2569

**Feats:**

- Added an example that uses loaders.gl to load .drc, .ply file formats as Playcanvas meshes.
- Added .drc and .ply file sources. These have been given to us under Creative Commons Attribution. The original source file can be found [here](https://sketchfab.com/3d-models/sample-6mm-73b432657d6d4d1aaec708d52d2d2d29). I used the [Google Draco Encoder](https://github.com/google/draco) to encode the .ply into .drc format for this example. 
 
Thanks to Draco compression, the original source .ply of 18.6mb, gets converted into a .drc of 2.8mb. The included .ply source is considered to be a small point cloud comprising of only 980000 points. It is not uncommon for point cloud files to go far above a million points. Here is where Draco mesh compression becomes a huge help in the transmission of these files across the network.

It is good to keep in mind that loaders.gl offers a range of other readers and writers which could be integrated into the engine similarly. Pointcloud formats like PCD and LAS aswell as 3D Tiles loaders for Potree + CesiumIon are some potentially useful formats. Aswell as this, loaders.gl also has a GLB/GLTF Writer which might help with the ability to export scenes from Playcanvas aswell.

The screenshot below shows a Draco compressed pointcloud mesh loaded into a Playcanvas scene with the help of loaders.gl. We are able to swap between .ply and .drc formats using either a double tap on mobile, or the left/right arrows on desktop.

image

This PR is related to comments in #2001 and gives a first example to show loaders.gl integrated with Playcanvas as requested in #2569.

@mvaligursky apologies this took me so long! Hopefully all is in order.

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

**Comment by mvaligursky** [comment]
Thanks, and no rush, take as time as needed @tonetechnician .
Could you please add a note on where the photogrammetry.drc and .ply files were sourced and what their distribution license is?

**Comment by mvaligursky** [comment]
Also, feel free to add a screenshot to the description .. these are always appreciated.

File: examples/loaders/loader-loaders.gl.html [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+        uniform mat4   matrix_view;
+
+        // Color to fragment program
+        varying vec4 outColor;
+
+
+        void main(void)
+        {
+            // Transform the geometry
+            mat4 modelView = matrix_view * matrix_model;
```
Comment (mvaligursky) [inline examples/loaders/loader-loaders.gl.html:82]: modelView matrix is not used, this line + uniform handling should be removed?

File: examples/loaders/loader-loaders.gl.html [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+        }
+    </script>
+
+    <script id="fshader" type="x-shader/x-fragment">
+        precision mediump float;
+        varying vec4 outColor;
+
+        void main(void)
+        {
+            float r = 0.0, delta = 0.0, alpha = 1.0;
```
Comment (mvaligursky) [inline examples/loaders/loader-loaders.gl.html:98]: some unused variables here.

File: examples/loaders/loader-loaders.gl.html [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+
+        app.assets.loadFromUrl('../assets/scripts/camera/orbit-camera.js', 'script', function (err, asset) {
+            if (!err && asset) {
+                run();
+            }
+        });
+
+        // Create shader to render mesh as circular points with colour
+        var shaderDefinition = {
+        attributes: {
```
Comment (mvaligursky) [inline examples/loaders/loader-loaders.gl.html:147]: attributes section should be indented

**Comment by tonetechnician** [comment]
Thanks @mvaligursky! 

Yes, this model is from a friend. He has it uploaded to Sketchfab [here](https://sketchfab.com/3d-models/sample-6mm-73b432657d6d4d1aaec708d52d2d2d29) and is licensed under creative commons. Is there a specific place that I should put this license info?

**Comment by Maksims** [comment]
It will be good to keep example code within coding standards and style of other examples. So that developers can focus on actual subject of an example, without the need to re-adjust on coding style and composition.

**Comment by tonetechnician** [comment]
Thanks @Maksims . Are you referring to the controls to swap between the different meshes + the info display in the example? Or the use of async patterns? 

I've tried to follow the coding guide in [CONTRIBUTING.md](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md)

**Comment by Maksims** [comment]
@tonetechnician good change. ES6 async and promises - they are great patterns, but unfortunately they are not supported by all webgl supported browsers, even if market share is tiny, we try to cover all the web, from oldest mobile devices and desktops to the high-end up to date platforms.

**Comment by tonetechnician** [comment]
Cool, that's clear and makes sense to me @Maksims. Thanks. It would seem [loaders.gl](https://loaders.gl/docs) opt for ES6 syntax which obviously doesn't align with this. They do provide some polyfills though - I'll edit to see if I can incorporate their polyfills and swap out the async + promises stuff.

**Comment by mvaligursky** [comment]
@tonetechnician - can you please use attached point cloud instead to avoid licensing issue. I hope that works fine.
[PointCloudPark.zip](https://github.com/playcanvas/engine/files/5730988/PointCloudPark.zip)

**Comment by tonetechnician** [comment]
Hey @mvaligursky, sure. That's no problem. Thanks for converting to .drc aswell!

**Comment by willeastcott** [comment]
Any updates on plans for this @mvaligursky?

**Comment by tonetechnician** [comment]
@willeastcott  @mvaligursky 

From my side, the main limitation to get this complete is the compatibility with IE11. Everything works really well in the other browsers, but I have run into some hiccups to get it working smooth on IE11. 

Loaders.gl had an issue which forced me to transpile their cdn distributions to es5 which was obviously extremely slow. I've raised an issue with them and they are in the [process](https://github.com/visgl/loaders.gl/pull/1074) of getting their distribution to export es5 properly, at which point I'll be tackling this PR again. 

I will say that the IE11 implementation is likely still going to be slow even after removing my own transpilation since loaders.gl use shared workers which aren't supported in IE11 to my knowledge. This requires me to disable using workers which will slow the process down for any browser when they run the example, although I'm thinking to just add a flag to disable/enable the workers dependent on the browser. 

I will update as there is more feedback from loaders.gl team.

**Comment by mvaligursky** [comment]
@tonetechnician - that's some great effort you're going to with this, and that's much appreciated.

**Comment by mvaligursky** [comment]
To be honest, I'd be keen to get this in even without IE11 support - it would not be any different to those examples we have showing WebGl2 only functionality.

**Comment by tonetechnician** [comment]
Hey @mvaligursky

Yes, if it's possible that'd be great. Last I spoke with the loaders.gl guys they said we need to use polyfills --> https://github.com/visgl/loaders.gl/issues/1143#event-4274206334 

It seems they do not want to change the syntax of it's usage. We'd need to still use it, but with polyfill.

**Comment by mvaligursky** [comment]
closing, this example has been refactored into https://github.com/playcanvas/engine/pull/3256

**Pull request closed without merging**